### PR TITLE
Implement trampoline support for AArch64

### PR DESCRIPTION
**Pull request by hsoontie**
Changes to openj9\runtime\compiler\runtime\Trampoline.cpp to add
trampoline support for AArch64.

Signed-off-by: Hillary Soontiens <hsoontie@unb.ca>

### CODE REVIEW

**Comment by hsoontie** [comment]
The code under this PR has not been compiled or tested as of the time of this comment.
@0xdaryl , please let me know what you think when you have a moment. Thanks!

File: runtime/compiler/runtime/Trampoline.cpp [hunk 1]
```diff
@@ -858,6 +858,108 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
...
+   if (branchInstr != 0x94000000)
+      {
+      // This is not a 'bl' instruction -- Don't patch
+      return true;
+      }
+
+   if (TR::Options::getCmdLineOptions()->getOption(TR_StressTrampolines)
+            || distance>(intptrj_t)TR::Compiler->target.cpu.maxUnconditionalBranchImmediateForwardOffset() 
+            || distance<(intptrj_t)TR::Compiler->target.cpu.maxUnconditionalBranchImmediateBackwardOffset()
+   ){
```
Comment (0xdaryl) [inline runtime/compiler/runtime/Trampoline.cpp]: Fix brace formatting

File: runtime/compiler/runtime/Trampoline.cpp [hunk 2]
```diff
@@ -858,6 +858,108 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
...
+      if (currentPC == newAddrOfCallee)
+         {
+         newTramp = currentTramp; 
+         }
+      else
+         {
+         newTramp = mcc_replaceTrampoline(reinterpret_cast<TR_OpaqueMethodBlock *>(callee), callSite, currentTramp, currentPC, newAddrOfCallee, false);
+         TR_ASSERT_FATAL(newTramp != NULL, "This is an internal error.\n");
```
Comment (0xdaryl) [inline runtime/compiler/runtime/Trampoline.cpp]: Could you provide a more descriptive assert message when you can't replace an existing trampoline?

File: runtime/compiler/runtime/Trampoline.cpp [hunk 3]
```diff
@@ -858,6 +858,108 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
...
+   TR_LinkageInfo *linkInfo = TR_LinkageInfo::get(newAddrOfCallee);
+   uint8_t        *entryAddress = (uint8_t *)newAddrOfCallee + linkInfo->getReservedWord();
+   intptrj_t       distance;
+   int32_t         currentDistance;
+   int32_t         branchInstr = *(int32_t *)callSite;
+   void           *newTramp;
+
+   distance = entryAddress - (uint8_t *)callSite;
+   currentDistance = (branchInstr << 6) >> 4;
+   branchInstr &= 0xff000000;
```
Comment (0xdaryl) [inline runtime/compiler/runtime/Trampoline.cpp]: If the branch instruction is a `BL`, then shouldn't this mask be `0xfc00000` to just mask the top 6 bits?

File: runtime/compiler/runtime/Trampoline.cpp [hunk 4]
```diff
@@ -858,6 +858,108 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
...
+            *((uint32_t*)currentTramp+1) = (uint32_t)entryAddress;
+            }
+         }
+
+      distance = (uint8_t *)newTramp - (uint8_t *)callSite;
+      }
+
+   if (currentDistance != distance)
+      {
+      branchInstr |= (distance >> 2) & 0x00ffffff;
```
Comment (0xdaryl) [inline runtime/compiler/runtime/Trampoline.cpp]: The immediate field is 26 bits.  Hence, shouldn't this mask be `0x3ffffff` ?

**Comment by 0xdaryl** [comment]
And please update the copyright header on the file (hence the Copyright Check failure).

File: runtime/compiler/runtime/Trampoline.cpp (hunk 3 above)
Comment (hsoontie) [reply runtime/compiler/runtime/Trampoline.cpp]: Sorry, you're right! I forgot to change this and the other mask.

**Comment by hsoontie** [comment]
@0xdaryl Hmm, it seems like I'm unable to view the copyright check details (404 error). I have made all the other changes.

**Comment by hsoontie** [comment]
@0xdaryl Here is another update! It looks like the copyright fixed itself.

**Comment by 0xdaryl** [comment]
This is an AArch64-specific change.  A Jenkins build is pointless.

**Comment by hsoontie** [comment]
@0xdaryl Removed WIP from the title, it should be good now!

**Comment by knn-k** [comment]
I get the following error during building for AArch64 in the Docker cross-compiling environment after this PR is merged:
```
/root/openj9-openjdk-jdk11/build/linux-aarch64-normal-server-release/vm/compiler/../compiler/runtime/Trampoline.cpp: In function 'bool arm64CodePatching(void*, void*, void*, void*, void*, void*)':
/root/openj9-openjdk-jdk11/build/linux-aarch64-normal-server-release/vm/compiler/../compiler/runtime/Trampoline.cpp:933:54: error: cast from 'uint8_t* {aka unsigned char*}' to 'uint32_t {aka unsigned int}' loses precision [-fpermissive]
             *((uint32_t*)currentTramp+1) = (uint32_t)entryAddress;
                                                      ^~~~~~~~~~~~
```

**Comment by 0xdaryl** [comment]
@hsoontie , please investigate.

**Comment by hsoontie** [comment]
@0xdaryl @knn-k , as I mentioned above I cannot compile this code just yet, so if I tried to fix it now it would involve some guesswork. If you wanted to get rid of this error ASAP, then I think you may want to consider reversing the merge for now until I can clarify that it is 100% fixed. I'm sorry about that!

**Comment by 0xdaryl** [comment]
No problem.  But please do your best to fix.  @knn-k can test the fix locally to verify before we merge.

**Comment by knn-k** [comment]
Some problems were found with this code.  #6829 fixes them.

**Pull request merged**